<a href="https://colab.research.google.com/github/Joeaviator/CAPSTONE_OF_MODELS/blob/main/CAPSTONE_MODELS.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [25]:
import numpy as np
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.preprocessing import StandardScaler

In [5]:
from google.colab import files
file = files.upload()

Saving water_potability.csv to water_potability.csv


In [6]:
df=pd.read_csv('water_potability.csv')
df.head()

,ph,Hardness,Solids,Chloramines,Sulfate,Conductivity,Organic_carbon,Trihalomethanes,Turbidity,Potability
0,NaN,204.890456,20791.31898,7.300212,368.516441,564.308654,10.379783,86.990970,2.963135,0
1,3.716080,129.422921,18630.05786,6.635246,NaN,592.885359,15.180013,56.329076,4.500656,0
2,8.099124,224.236259,19909.54173,9.275884,NaN,418.606213,16.868637,66.420093,3.055934,0
3,8.316766,214.373394,22018.41744,8.059332,356.886136,363.266516,18.436525,100.341674,4.628771,0
4,9.092223,181.101509,17978.98634,6.546600,310.135738,398.410813,11.558279,31.997993,4.075075,0


In [33]:
df.duplicated().sum()

np.int64(0)

In [7]:
df.isnull().sum()

,0
ph,491
Hardness,0
Solids,0
Chloramines,0
Sulfate,781
Conductivity,0
Organic_carbon,0
Trihalomethanes,162
Turbidity,0
Potability,0


In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3276 entries, 0 to 3275
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   ph               2785 non-null   float64
 1   Hardness         3276 non-null   float64
 2   Solids           3276 non-null   float64
 3   Chloramines      3276 non-null   float64
 4   Sulfate          2495 non-null   float64
 5   Conductivity     3276 non-null   float64
 6   Organic_carbon   3276 non-null   float64
 7   Trihalomethanes  3114 non-null   float64
 8   Turbidity        3276 non-null   float64
 9   Potability       3276 non-null   int64  
dtypes: float64(9), int64(1)
memory usage: 256.1 KB


In [11]:
df=df.fillna(0.0)

In [12]:
df.isnull().sum()

,0
ph,0
Hardness,0
Solids,0
Chloramines,0
Sulfate,0
Conductivity,0
Organic_carbon,0
Trihalomethanes,0
Turbidity,0
Potability,0


In [18]:
#Feature scale
x=df.drop('Potability', axis=1)
y=df['Potability']

In [17]:
x.head()

,ph,Hardness,Solids,Chloramines,Sulfate,Conductivity,Organic_carbon,Trihalomethanes,Turbidity
0,0.000000,204.890456,20791.31898,7.300212,368.516441,564.308654,10.379783,86.990970,2.963135
1,3.716080,129.422921,18630.05786,6.635246,0.000000,592.885359,15.180013,56.329076,4.500656
2,8.099124,224.236259,19909.54173,9.275884,0.000000,418.606213,16.868637,66.420093,3.055934
3,8.316766,214.373394,22018.41744,8.059332,356.886136,363.266516,18.436525,100.341674,4.628771
4,9.092223,181.101509,17978.98634,6.546600,310.135738,398.410813,11.558279,31.997993,4.075075


In [19]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=42)

In [26]:
scale=StandardScaler()
x_train_scale=scale.fit_transform(x_train)
x_test_scale=scale.fit_transform(x_test)

In [29]:
model=LogisticRegression(max_iter=1500)
model.fit(x_train_scale,y_train)
preds=model.predict(x_test_scale)
accuracy_score(preds,y_test)

0.6280487804878049

#Accuracy Score: 0.63

In [30]:
Knn=KNeighborsClassifier(n_neighbors=5)
Knn.fit(x_train_scale,y_train)
preds=Knn.predict(x_test_scale)
accuracy_score(preds,y_test)

0.5701219512195121

#Accuracy Score: 0.570

In [31]:
model_2=DecisionTreeClassifier()
model_2.fit(x_train,y_train)
preds=model_2.predict(x_test)
accuracy_score(preds,y_test)

0.573170731707317

#Accuracy Score: 0.573



In [32]:
model_3=RandomForestClassifier()
model_3.fit(x_train,y_train)
preds=model_3.predict(x_test)
accuracy_score(preds,y_test)

0.6890243902439024

#Accuracy Score: 0.69

#Findings
* The best model is the Random Forest Classifier
* The dataset is diifcult for models to predict
* It requires two different variations of the same data to maintain a respectable prediction rate
* Overall the missing null values seem to have impacted the model's ability to predict fluently and accurately